letsencrypt/acme/challenges.py

```diff
@@ -0,0 +1,360 @@
+        :type challenge: :class:`DVSNI`
+
+        """
+        z = hashlib.new("sha256")  # pylint: disable=invalid-name
+        z.update(chall.r)
+        z.update(self.s)
+        return z.hexdigest()
+
+    def z_domain(self, chall):
+        """Domain name for certificate subjectAltName."""
+        return self.z(chall) + self.DOMAIN_SUFFIX
+
+    def _fields_to_json(self):
+        return {"s": jose.b64encode(self.s)}
+
+    @classmethod
+    def from_valid_json(cls, jobj):
+        return cls(s=cls._decode_b64jose(jobj["s"], cls.S_SIZE))
+
+
+@Challenge.register
+class RecoveryContact(ClientChallenge):
+    """ACME "recoveryContact" challenge."""
+    acme_type = "recoveryContact"
+    __slots__ = ("activation_url", "success_url", "contact")
+
+    def _fields_to_json(self):
+        fields = {}
+        add = functools.partial(_extend_if_not_none, fields)
+        add(self.activation_url, "activationURL")
+        add(self.success_url, "successURL")
+        add(self.contact, "contact")
+        return fields
+
+    @classmethod
+    def from_valid_json(cls, jobj):
+        return cls(activation_url=jobj.get("activationURL"),
+                   success_url=jobj.get("successURL"),
+                   contact=jobj.get("contact"))
+
+
+@ChallengeResponse.register
+class RecoveryContactResponse(ChallengeResponse):
+    """ACME "recoveryContact" challenge response."""
+    acme_type = "recoveryContact"
+    __slots__ = ("token",)
+
+    def _fields_to_json(self):
+        fields = {}
+        if self.token is not None:
+            fields["token"] = self.token
+        return fields
+
+    @classmethod
+    def from_valid_json(cls, jobj):
+        return cls(token=jobj.get("token"))
+
+
+@Challenge.register
+class RecoveryToken(ClientChallenge):
+    """ACME "recoveryToken" challenge."""
+    acme_type = "recoveryToken"
+    __slots__ = ()
+
+    def _fields_to_json(self):
+        return {}
+
+    @classmethod
+    def from_valid_json(cls, jobj):
+        return cls()
+
+
+@ChallengeResponse.register
+class RecoveryTokenResponse(ChallengeResponse):
+    """ACME "recoveryToken" challenge response."""
+    acme_type = "recoveryToken"
+    __slots__ = ("token",)
+
+    def _fields_to_json(self):
+        fields = {}
+        if self.token is not None:
+            fields["token"] = self.token
+        return fields
+
+    @classmethod
+    def from_valid_json(cls, jobj):
+        return cls(token=jobj.get("token"))
+
+
+def _extend_if_not_empty(dikt, param, name):
+    if param:
+        dikt[name] = param
+
+def _extend_if_not_none(dikt, param, name):
+    if param is not None:
+        dikt[name] = param
+
+
+@Challenge.register
+class ProofOfPossession(ClientChallenge):
+    """ACME "proofOfPossession" challenge.
+
+    :ivar str nonce: Random data, **not** base64-encoded.
+    :ivar hints: Various clues for the client (:class:`Hints`).
+
+    """
+    acme_type = "proofOfPossession"
+    __slots__ = ("alg", "nonce", "hints")
```

Comment:
For the time being I follow the policy that only mangled fields are documented (to reduce copy&pasting from acme-spec). If we do decide to include non-mangled fields as well,  this will have to be done for entire `acme` module. Is that OK with you if I leave it as is for now?